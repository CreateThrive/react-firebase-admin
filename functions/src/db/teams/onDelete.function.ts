import { database } from 'firebase-admin';
import { database as db } from 'firebase-functions';
import { BatchUpdate, IUserRtdb } from '../../types';

export default db.ref('teams/{teamId}').onDelete(async (snapshot, context) => {
  const { teamId } = context.params;
  let users;

  try {
    users = (await database().ref('users').once('value')).val();
  } catch (error) {
    console.log(error);
  }

  users = Object.entries(users).filter(([key, user]) => {
    const { teams } = user as IUserRtdb;

    return teams?.includes(teamId);
  });

  const batchUpdate: BatchUpdate = {};

  users.forEach(([key, user]) => {
    const { teams } = user as IUserRtdb;

    if (teams) {
      const newTeams = teams.filter((index: string) => index !== teamId);

      batchUpdate[`/${key}/teams`] = newTeams;
    }
  });

  return database().ref('users').update(batchUpdate);
});
