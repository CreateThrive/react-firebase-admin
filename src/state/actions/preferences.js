import { createAction } from 'redux-act';

export const PREFERENCES_SET_LOCALE = createAction('PREFERENCES_SET_LOCALE');

export const setUserLocale = locale => dispatch => {
  return dispatch(PREFERENCES_SET_LOCALE({ locale }));
};
