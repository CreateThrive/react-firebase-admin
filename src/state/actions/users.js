import { createAction } from 'redux-act';
import { toastr } from 'react-redux-toastr';

import { firebaseError } from 'utils';
import firebase from 'firebase.js';
import { checkUserData, AUTH_UPDATE_USER_DATA } from './auth';

export const USERS_FETCH_DATA_INIT = createAction('USERS_FETCH_DATA_INIT');
export const USERS_FETCH_DATA_SUCCESS = createAction(
  'USERS_FETCH_DATA_SUCCESS'
);
export const USERS_FETCH_DATA_FAIL = createAction('USERS_FETCH_DATA_FAIL');

export const USERS_DELETE_USER_INIT = createAction('USERS_DELETE_USER_INIT');
export const USERS_DELETE_USER_SUCCESS = createAction(
  'USERS_DELETE_USER_SUCCESS'
);
export const USERS_DELETE_USER_FAIL = createAction('USERS_DELETE_USER_FAIL');

export const USERS_CLEAR_DATA = createAction('USERS_CLEAR_DATA');

export const USERS_CREATE_USER_INIT = createAction('USERS_CREATE_USER_INIT');
export const USERS_CREATE_USER_SUCCESS = createAction(
  'USERS_CREATE_USER_SUCCESS'
);
export const USERS_CREATE_USER_FAIL = createAction('USERS_CREATE_USER_FAIL');

export const USERS_MODIFY_USER_INIT = createAction('USERS_MODIFY_USER_INIT');
export const USERS_MODIFY_USER_SUCCESS = createAction(
  'USERS_MODIFY_USER_SUCCESS'
);
export const USERS_MODIFY_USER_FAIL = createAction('USERS_MODIFY_USER_FAIL');

export const USERS_CLEAN_UP = createAction('USERS_CLEAN_UP');

export const USERS_CLEAR_DATA_LOGOUT = createAction('USERS_CLEAR_DATA_LOGOUT');

export const fetchUsers = () => {
  return async (dispatch, getState) => {
    dispatch(checkUserData());

    dispatch(USERS_FETCH_DATA_INIT());

    const { id } = getState().auth.userData;

    let users;

    try {
      users = (
        await firebase
          .database()
          .ref('users')
          .once('value')
      ).val();
    } catch (error) {
      toastr.error('', error);
      return dispatch(USERS_FETCH_DATA_FAIL({ error }));
    }

    const usersData = users
      ? Object.entries(users).map(([key, value]) => ({
          id: key,
          ...value
        }))
      : [];

    return dispatch(
      USERS_FETCH_DATA_SUCCESS({
        users: usersData.filter(user => user.id !== id)
      })
    );
  };
};

const deleteLogo = oldLogo => {
  if (!oldLogo.includes('firebasestorage')) {
    return null;
  }
  const logoPath = oldLogo
    .split('users%2F')
    .pop()
    .split('?alt=media')
    .shift();
  return firebase
    .storage()
    .ref(`users/${logoPath}`)
    .delete();
};

export const deleteUser = id => {
  return async (dispatch, getState) => {
    dispatch(USERS_DELETE_USER_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState()
      .users.data.filter(user => user.id === id)
      .pop();

    const deleteLogoTask = logoUrl ? deleteLogo(logoUrl) : null;

    const deleteUserTask = firebase
      .database()
      .ref(`users/${id}`)
      .remove();

    try {
      await Promise.all([deleteLogoTask, deleteUserTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        USERS_DELETE_USER_FAIL({
          error: errorMessage
        })
      );
    }

    toastr.success('', 'The user was deleted.');
    return dispatch(USERS_DELETE_USER_SUCCESS({ id }));
  };
};

export const clearUsersData = () => {
  return dispatch => {
    dispatch(USERS_CLEAR_DATA());
  };
};

export const clearUsersDataLogout = () => {
  return dispatch => {
    dispatch(USERS_CLEAR_DATA_LOGOUT());
  };
};

const uploadLogo = (uid, file) => {
  const storageRef = firebase.storage().ref();

  const fileExtension = file.name.split('.').pop();

  const fileName = `${uid}.${fileExtension}`;

  return storageRef.child(`users/${fileName}`).put(file);
};

const getLogoUrl = (uid, file) => {
  const fileExtension = file.name.split('.').pop();

  const bucketUrl = `${process.env.REACT_APP_FIRE_BASE_STORAGE_API}`;

  return `${bucketUrl}/o/users%2F${uid}_200x200.${fileExtension}?alt=media`;
};

export const createUser = ({
  name,
  email,
  location,
  file,
  createdAt,
  isAdmin
}) => {
  return async (dispatch, getState) => {
    dispatch(USERS_CREATE_USER_INIT());
    const { locale } = getState().preferences;

    let response;
    try {
      const createUserAuth = firebase
        .functions()
        .httpsCallable('httpsCreateUser');

      response = await createUserAuth({ email, isAdmin });
    } catch (error) {
      const errorMessage = firebaseError(error.message, locale);
      toastr.error('', errorMessage);
      return dispatch(
        USERS_CREATE_USER_FAIL({
          error: errorMessage
        })
      );
    }

    const { uid } = response.data;

    let uploadLogoTask = null;
    let logoUrl = null;
    if (file) {
      logoUrl = getLogoUrl(uid, file);
      uploadLogoTask = uploadLogo(uid, file);
    }

    const createUserDbTask = firebase
      .database()
      .ref(`users/${uid}`)
      .set({ name, email, location, logoUrl, createdAt, isAdmin });

    const actionCodeSettings = {
      url: process.env.REACT_APP_LOGIN_PAGE_URL,
      handleCodeInApp: true
    };

    const sendSignInLinkToEmailTask = firebase
      .auth()
      .sendSignInLinkToEmail(email, actionCodeSettings);

    try {
      await Promise.all([
        uploadLogoTask,
        createUserDbTask,
        sendSignInLinkToEmailTask
      ]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        USERS_CREATE_USER_FAIL({
          error: errorMessage
        })
      );
    }

    toastr.success('', 'User created successfully');
    return dispatch(USERS_CREATE_USER_SUCCESS({ user: response.data }));
  };
};

export const modifyUser = ({
  name,
  location,
  isAdmin,
  file,
  createdAt,
  id,
  isEditing,
  isProfile
}) => {
  return async (dispatch, getState) => {
    dispatch(USERS_MODIFY_USER_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = isProfile
      ? getState().auth.userData
      : getState()
          .users.data.filter(user => user.id === id)
          .pop();

    let deleteLogoTask;
    let uploadLogoTask;
    let newLogoUrl = null;
    if (file) {
      newLogoUrl = getLogoUrl(id, file);
      deleteLogoTask = logoUrl && deleteLogo(logoUrl);
      uploadLogoTask = uploadLogo(id, file);
    }

    const userData = {
      name,
      location,
      createdAt,
      isAdmin,
      logoUrl: logoUrl || newLogoUrl
    };

    const updateUserDbTask = firebase
      .database()
      .ref(`users/${id}`)
      .update(userData);

    const { uid } = firebase.auth().currentUser;

    if (id === uid) {
      dispatch(AUTH_UPDATE_USER_DATA({ ...userData, id }));
    }

    try {
      await Promise.all([deleteLogoTask, uploadLogoTask, updateUserDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        USERS_MODIFY_USER_FAIL({
          error: errorMessage
        })
      );
    }

    if (isProfile) {
      toastr.success('', 'Profile updated successfully');
    } else if (isEditing) {
      toastr.success('', 'User updated successfully');
    }

    return dispatch(USERS_MODIFY_USER_SUCCESS({ user: { ...userData, id } }));
  };
};

export const usersCleanUp = () => dispatch => dispatch(USERS_CLEAN_UP());
