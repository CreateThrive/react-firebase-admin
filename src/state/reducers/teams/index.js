import { createReducer } from 'redux-act';

import {
  TEAMS_FETCH_TEAMS_INIT,
  TEAMS_FETCH_TEAMS_SUCCESS,
  TEAMS_FETCH_TEAMS_FAIL,
  TEAMS_DELETE_TEAM_INIT,
  TEAMS_DELETE_TEAM_SUCCESS,
  TEAMS_DELETE_TEAM_FAIL,
  TEAMS_CREATE_TEAM_INIT,
  TEAMS_CREATE_TEAM_SUCCESS,
  TEAMS_CREATE_TEAM_FAIL,
  TEAMS_MODIFY_TEAM_INIT,
  TEAMS_MODIFY_TEAM_SUCCESS,
  TEAMS_MODIFY_TEAM_FAIL,
  TEAMS_CLEAN_UP,
  TEAMS_ADD_USER_FAIL,
  TEAMS_ADD_USER_INIT,
  TEAMS_ADD_USER_SUCCESS,
  TEAMS_REMOVE_USER_FAIL,
  TEAMS_REMOVE_USER_INIT,
  TEAMS_REMOVE_USER_SUCCESS,
  TEAMS_SELECTED_TEAM,
  TEAMS_FETCH_TEAM_USERS_FAIL,
  TEAMS_FETCH_TEAM_USERS_INIT,
  TEAMS_FETCH_TEAM_USERS_SUCCESS,
} from 'state/actions/teams';

const initialState = {
  teamsList: [],
  usersList: [],
  loadingTeams: false,
  loadingUsers: false,
  error: null,
  success: false,
  deleted: false,
  teamId: 0,
};

export const teamsReducer = createReducer(
  {
    [TEAMS_FETCH_TEAMS_INIT]: () => ({
      ...initialState,
      loadingTeams: true,
    }),
    [TEAMS_FETCH_TEAMS_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: payload.teams,
      loadingTeams: false,
      error: null,
    }),
    [TEAMS_FETCH_TEAMS_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_DELETE_TEAM_INIT]: (state) => ({
      ...state,
      loadingTeams: true,
    }),
    [TEAMS_DELETE_TEAM_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: state.teamsList.filter((elem) => elem.id !== payload.id),
      loadingTeams: false,
      error: null,
      deleted: true,
    }),
    [TEAMS_DELETE_TEAM_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_CREATE_TEAM_INIT]: (state) => ({
      ...state,
      loadingTeams: true,
    }),
    [TEAMS_CREATE_TEAM_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: state.teamsList.concat(payload.team),
      loadingTeams: false,
      error: null,
      success: true,
    }),
    [TEAMS_CREATE_TEAM_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_MODIFY_TEAM_INIT]: (state) => ({
      ...state,
      loadingTeams: true,
    }),
    [TEAMS_MODIFY_TEAM_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: !state.teamsList
        ? []
        : state.teamsList.map((elem) => {
            if (elem.id === payload.id) {
              return {
                ...elem,
                name: payload.team.name,
                description: payload.team.description,
                logoUrl: payload.team.logoUrl || null,
              };
            }
            return elem;
          }),
      loadingTeams: false,
      error: null,
      success: true,
    }),
    [TEAMS_MODIFY_TEAM_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_CLEAN_UP]: (state) => ({
      ...state,
      loadingTeams: false,
      error: null,
      success: false,
      deleted: false,
    }),
    [TEAMS_ADD_USER_FAIL]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: payload.error,
    }),
    [TEAMS_ADD_USER_INIT]: (state) => ({
      ...state,
      loadingUsers: true,
    }),
    [TEAMS_ADD_USER_SUCCESS]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: null,
      success: true,
      teamsList: state.teamsList.map((value) => {
        if (value.id === payload.teamId) {
          return { ...value, users: payload.users };
        }
        return value;
      }),
      usersList: [...state.usersList, payload.user],
    }),
    [TEAMS_REMOVE_USER_FAIL]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: payload.error,
    }),
    [TEAMS_REMOVE_USER_INIT]: (state) => ({
      ...state,
      loadingUsers: true,
    }),
    [TEAMS_REMOVE_USER_SUCCESS]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: null,
      teamsList: state.teamsList.map((value) => {
        if (value.id === payload.teamId) {
          return { ...value, users: payload.users };
        }
        return value;
      }),
      usersList: state.usersList.filter((value) => value.id !== payload.userId),
      deleted: true,
    }),
    [TEAMS_SELECTED_TEAM]: (state, payload) => ({
      ...state,
      teamId: payload.teamId,
    }),
    [TEAMS_FETCH_TEAM_USERS_INIT]: (state) => ({
      ...state,
      loadingUsers: true,
    }),
    [TEAMS_FETCH_TEAM_USERS_SUCCESS]: (state, payload) => ({
      ...state,
      usersList: payload.users || [],
      loadingUsers: false,
      error: null,
    }),
    [TEAMS_FETCH_TEAM_USERS_FAIL]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: payload.error,
    }),
  },
  initialState
);
