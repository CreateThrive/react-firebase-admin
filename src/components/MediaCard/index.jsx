/* eslint-disable jsx-a11y/control-has-associated-label */
import React from 'react';
import { Link } from 'react-router-dom';
import { useSelector, shallowEqual, useDispatch } from 'react-redux';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import { fetchTeamUsers } from 'state/actions/teams';
import { useFormatDate } from 'hooks';
import userDefaultLogo from 'assets/user-default-log.jpg';
import teamDeaultLogo from 'assets/team-default-log.png';

const MediaCard = ({ content, onDeleteHandler, onModifyHandler, isUser }) => {
  const { isAdmin, teamId } = useSelector(
    (state) => ({
      isAdmin: state.auth.userData.isAdmin,
      teamId: state.teams.teamId,
    }),
    shallowEqual
  );

  const dispatch = useDispatch();

  const onSelectedHandler = () => {
    if (!isUser && content.id !== teamId) {
      dispatch(fetchTeamUsers(content.id));
    }
  };

  const userCreatedAt = useFormatDate(content.createdAt, {
    weekday: 'short',
    year: 'numeric',
    month: 'short',
    day: 'numeric',
  });

  return (
    <div
      onClick={onSelectedHandler}
      className={classNames('media', {
        'has-background-light': content.id === teamId && !isUser,
      })}
      style={{
        marginTop: 'auto',
        marginBottom: 'auto',
        paddingBottom: '1rem',
        cursor: !isUser && content.id !== teamId && 'pointer',
      }}
      data-testid="media"
    >
      <figure className="media-left">
        <p className="image is-64x64">
          <img
            className="is-rounded"
            src={content.logoUrl || (isUser ? userDefaultLogo : teamDeaultLogo)}
            alt="Profile logo preview"
          />
        </p>
      </figure>
      <div className="media-content">
        <div className="content">
          <p className="media-meta">
            <strong>{content.name}</strong>{' '}
            {isUser && <small className="has-text-grey">{userCreatedAt}</small>}
          </p>
          {isUser ? <p>{content.email}</p> : <p>{content.description}</p>}
        </div>
      </div>

      {isAdmin && (
        <div className="media-right">
          <div className="field is-grouped is-grouped-multiline">
            <div className="control">
              {isUser ? (
                <Link to={`/users/${content.id}`}>
                  <span className="icon">
                    <i className="mdi mdi-account-edit" />
                  </span>
                </Link>
              ) : (
                <span
                  className="icon has-text-info"
                  onClick={onModifyHandler}
                  data-testid="modify"
                  style={{ cursor: 'pointer' }}
                >
                  <i className="mdi mdi-pencil" />
                </span>
              )}
            </div>
            <div className="control">
              <span
                className="icon has-text-danger"
                onClick={() => onDeleteHandler(content.id)}
                data-testid="delete"
                style={{ cursor: 'pointer' }}
              >
                <i className="mdi mdi-cancel" />
              </span>
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

const userPropTypes = PropTypes.shape({
  id: PropTypes.string.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  name: PropTypes.string.isRequired,
  location: PropTypes.string,
  logoUrl: PropTypes.string,
  createdAt: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
});

const teamPropTypes = PropTypes.shape({
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string,
  logoUrl: PropTypes.string,
  createdBy: PropTypes.string.isRequired,
});

MediaCard.propTypes = {
  isUser: PropTypes.bool,
  content: PropTypes.oneOfType([teamPropTypes, userPropTypes]).isRequired,
  onModifyHandler: PropTypes.func,
  onDeleteHandler: PropTypes.func.isRequired,
};

MediaCard.defaultProps = {
  isUser: false,
};

export default MediaCard;
