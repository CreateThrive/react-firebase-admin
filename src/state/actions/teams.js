import { createAction } from 'redux-act';
import { toastr } from 'react-redux-toastr';
import { v4 as uuid } from 'uuid';

import { firebaseError, deleteLogo, getLogoUrl, uploadLogo } from 'utils';
import {
  fetchCollection,
  createDocument,
  deleteDocument,
  updateDocument,
  batchUpdateDocument,
} from '../api';
import { removeStoreUser, addStoreUser } from './users';

export const TEAMS_FETCH_TEAMS_INIT = createAction('TEAMS_FETCH_TEAMS_INIT');

export const TEAMS_FETCH_TEAMS_SUCCESS = createAction(
  'TEAMS_FETCH_TEAMS_SUCCESS'
);

export const TEAMS_FETCH_TEAMS_FAIL = createAction('TEAMS_FETCH_TEAMS_FAIL');

export const TEAMS_DELETE_TEAM_INIT = createAction('TEAMS_DELETE_TEAM_INIT');

export const TEAMS_DELETE_TEAM_SUCCESS = createAction(
  'TEAMS_DELETE_TEAM_SUCCESS'
);

export const TEAMS_DELETE_TEAM_FAIL = createAction('TEAMS_DELETE_TEAM_FAIL');

export const TEAMS_CREATE_TEAM_INIT = createAction('TEAMS_CREATE_TEAM_INIT');

export const TEAMS_CREATE_TEAM_SUCCESS = createAction(
  'TEAMS_CREATE_TEAM_SUCCESS'
);
export const TEAMS_CREATE_TEAM_FAIL = createAction('TEAMS_CREATE_TEAM_FAIL');

export const TEAMS_MODIFY_TEAM_INIT = createAction('TEAMS_MODIFY_TEAM_INIT');

export const TEAMS_MODIFY_TEAM_SUCCESS = createAction(
  'TEAMS_MODIFY_TEAM_SUCCESS'
);
export const TEAMS_MODIFY_TEAM_FAIL = createAction('TEAMS_MODIFY_TEAM_FAIL');

export const TEAMS_CLEAN_UP = createAction('TEAMS_CLEAN_UP');

export const TEAMS_ADD_USER_INIT = createAction('TEAMS_ADD_USER_INIT');

export const TEAMS_ADD_USER_SUCCESS = createAction('TEAMS_ADD_USER_SUCCESS');

export const TEAMS_ADD_USER_FAIL = createAction('TEAMS_ADD_USER_FAIL');

export const TEAMS_REMOVE_USER_INIT = createAction('TEAMS_REMOVE_USER_INIT');

export const TEAMS_REMOVE_USER_SUCCESS = createAction(
  'TEAMS_REMOVE_USER_SUCCESS'
);

export const TEAMS_REMOVE_USER_FAIL = createAction('TEAMS_REMOVE_USER_FAIL');

export const TEAMS_SELECTED_TEAM = createAction('TEAMS_SELECTED_TEAM');

export const TEAMS_FETCH_TEAM_USERS_INIT = createAction(
  'TEAMS_FETCH_TEAM_USERS_INIT'
);

export const TEAMS_FETCH_TEAM_USERS_SUCCESS = createAction(
  'TEAMS_FETCH_TEAM_USERS_SUCCESS'
);

export const TEAMS_FETCH_TEAM_USERS_FAIL = createAction(
  'TEAMS_FETCH_TEAM_USERS_FAIL'
);

export const fetchTeams = () => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_FETCH_TEAMS_INIT());
    const { locale } = getState().preferences;
    const { isAdmin, id } = getState().auth.userData;

    let teams;
    try {
      teams = await fetchCollection(
        'teams',
        isAdmin
          ? {
              filterBy: 'users',
              value: id,
              isArray: true,
            }
          : {}
      );
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_FETCH_TEAMS_FAIL({ error }));
    }

    return dispatch(
      TEAMS_FETCH_TEAMS_SUCCESS({
        teams: isAdmin
          ? teams
          : teams.filter((value) => value.users?.includes(id)),
      })
    );
  };
};

export const deleteTeam = (id) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_DELETE_TEAM_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState().teams.teamsList.find(
      (team) => team.id === id
    );

    const deleteLogoTask = logoUrl ? deleteLogo(logoUrl, 'teams') : null;

    const deleteTeamTask = deleteDocument('teams', id);

    try {
      await Promise.all([deleteLogoTask, deleteTeamTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_DELETE_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'The team was deleted.');
    return dispatch(TEAMS_DELETE_TEAM_SUCCESS({ id }));
  };
};

export const createTeam = ({ name, description, file }) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_CREATE_TEAM_INIT());
    const { locale } = getState().preferences;

    const uid = uuid();

    let uploadLogoTask = null;
    let logoUrl = null;
    if (file) {
      logoUrl = getLogoUrl(uid, file, 'teams');
      uploadLogoTask = uploadLogo(uid, file, 'teams');
    }

    const teamData = {
      name,
      description,
      logoUrl,
      createdAt: new Date().toString(),
      createdBy: getState().auth.userData.name,
    };

    const createTeamDbTask = createDocument('teams', uid, teamData);

    try {
      await Promise.all([uploadLogoTask, createTeamDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_CREATE_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'Team created successfully');
    return dispatch(
      TEAMS_CREATE_TEAM_SUCCESS({ team: { ...teamData, id: uid } })
    );
  };
};

export const modifyTeam = ({ name, description, file, id }) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_MODIFY_TEAM_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState().teams.teamsList.find(
      (team) => team.id === id
    );

    let deleteLogoTask;
    let uploadLogoTask;
    let newLogoUrl = null;
    if (file) {
      newLogoUrl = getLogoUrl(id, file, 'teams');
      deleteLogoTask = logoUrl && deleteLogo(logoUrl, 'teams');
      uploadLogoTask = uploadLogo(id, file, 'teams');
    }

    const teamData = {
      name,
      description,
      logoUrl: logoUrl || newLogoUrl,
    };
    const updateTeamDbTask = updateDocument('teams', id, teamData);

    try {
      await Promise.all([deleteLogoTask, uploadLogoTask, updateTeamDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_MODIFY_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'Team updated successfully');

    return dispatch(TEAMS_MODIFY_TEAM_SUCCESS({ team: teamData, id }));
  };
};

export const teamsCleanUp = () => (dispatch) => dispatch(TEAMS_CLEAN_UP());

export const addUser = (teamId, user) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_ADD_USER_INIT());
    const { locale } = getState().preferences;
    const { teamsList } = getState().teams;

    const userId = user.id;

    const teamUsers =
      teamsList.find((value) => value.id === teamId)?.users || [];
    const userTeams = user?.teams || [];

    const newUserTeams = [...userTeams, teamId];
    const newTeamUsers = [...teamUsers, userId];

    const batchCollections = {
      users: {
        documents: [
          {
            id: userId,
            field: 'teams',
            value: newUserTeams,
          },
        ],
      },

      teams: {
        documents: [
          {
            id: teamId,
            field: 'users',
            value: newTeamUsers,
          },
        ],
      },
    };

    try {
      await batchUpdateDocument(batchCollections);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_ADD_USER_FAIL({ error }));
    }

    dispatch(removeStoreUser(userId));

    return dispatch(
      TEAMS_ADD_USER_SUCCESS({
        teamId,
        users: newTeamUsers,
        user: { ...user, teams: newUserTeams },
      })
    );
  };
};

export const removeUser = (teamId, userId) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_REMOVE_USER_INIT());
    const { locale } = getState().preferences;
    const { teamsList, usersList } = getState().teams;

    const user = usersList.find((value) => value.id === userId);
    const teamUsers =
      teamsList.find((value) => value.id === teamId)?.users || [];
    const userTeams = user?.teams || [];

    const newUserTeams = userTeams.filter((value) => value !== teamId);
    const newTeamUsers = teamUsers.filter((value) => value !== userId);

    const bulkUpdate = {};
    bulkUpdate[`/users/${userId}/teams`] = newUserTeams;
    bulkUpdate[`/teams/${teamId}/users`] = newTeamUsers;

    try {
      await updateDocument('', '', bulkUpdate);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_REMOVE_USER_FAIL({ error }));
    }

    dispatch(addStoreUser({ ...user, teams: newUserTeams }));

    return dispatch(
      TEAMS_REMOVE_USER_SUCCESS({
        teamId,
        users: newTeamUsers,
        userId,
      })
    );
  };
};

export const fetchTeamUsers = (teamId) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_FETCH_TEAM_USERS_INIT());
    const { locale } = getState().preferences;

    let users;
    try {
      users = await fetchCollection('users', {
        filterBy: 'users',
        value: teamId,
        isArray: true,
      });
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_FETCH_TEAM_USERS_FAIL({ error }));
    }

    dispatch(TEAMS_SELECTED_TEAM({ teamId }));

    return dispatch(
      TEAMS_FETCH_TEAM_USERS_SUCCESS({
        users: users.filter((value) => value.teams?.includes(teamId)),
      })
    );
  };
};
