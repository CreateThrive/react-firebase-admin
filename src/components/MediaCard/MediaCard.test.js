import { fireEvent } from '@testing-library/react';
import React from 'react';
import * as reactRedux from 'react-redux';

import * as actions from 'state/actions/teams';
import MediaCard from '.';

const onDelete = jest.fn();
const onModify = jest.fn();

const testTeam = { id: 'teamTestId', name: 'testTeam', createdBy: 'test' };
describe('<MediaCard /> rendering', () => {
  const dispatchMock = jest.fn();
  let userData;

  beforeEach(() => {
    userData = {
      email: 'mkrukuy@gmail.com',
      name: 'Mateo',
      location: 'Montevideo, Uruguay',
      isAdmin: true,
      file: null,
      id: 'testId',
      logoUrl: 'some logoUrl',
      createdAt: '11/12/2020',
    };
    jest
      .spyOn(reactRedux, 'useDispatch')
      .mockImplementation(() => dispatchMock);
    jest.spyOn(actions, 'fetchTeamUsers').mockImplementation(jest.fn);
  });

  it('should render without crashing', () => {
    const { component } = renderWithProviders(
      <MediaCard content={userData} onDeleteHandler={onDelete} isUser />
    )({
      auth: { userData },
      teams: { teamid: 'testId' },
    });

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should dispatch fetchTeamUsers action when clicking the component', () => {
    const { component } = renderWithProviders(
      <MediaCard content={testTeam} onDeleteHandler={onDelete} />
    )({
      auth: { userData },
      teams: { teamid: 'testId' },
    });

    fireEvent.click(component.getByTestId('media'));

    expect(actions.fetchTeamUsers).toBeCalledTimes(1);

    expect(actions.fetchTeamUsers).toBeCalledWith('teamTestId');
  });

  it('should should not render the delete button for non-admin users', () => {
    const { component } = renderWithProviders(
      <MediaCard
        content={testTeam}
        onDeleteHandler={onDelete}
        onModifyHandler={onModify}
      />
    )({
      auth: { userData: { ...userData, isAdmin: false } },
      teams: { teamid: 'teamTestId' },
    });

    expect(component.queryByTestId('delete')).toBeNull();
  });

  it('should dispatch onDelete action when clicking the delete button', () => {
    const { component } = renderWithProviders(
      <MediaCard
        content={testTeam}
        onDeleteHandler={onDelete}
        onModifyHandler={onModify}
      />
    )({
      auth: { userData },
      teams: { teamid: 'teamTestId' },
    });

    fireEvent.click(component.getByTestId('delete'));

    expect(onDelete).toBeCalledTimes(1);
  });

  it('should dispatch onModify action when clicking the modify button', () => {
    const { component } = renderWithProviders(
      <MediaCard
        content={userData}
        onDeleteHandler={onDelete}
        onModifyHandler={onModify}
      />
    )({
      auth: { userData },
      teams: { teamid: 'teamTestId' },
    });

    fireEvent.click(component.getByTestId('modify'));

    expect(onModify).toBeCalledTimes(1);
  });

  it('should redirect to the user profile when clicking the modify button', () => {
    const { component } = renderWithProviders(
      <MediaCard
        content={userData}
        onDeleteHandler={onDelete}
        onModifyHandler={onModify}
        isUser
      />
    )({
      auth: { userData },
      teams: { teamid: 'teamTestId' },
    });

    fireEvent.click(component.getByRole('link'));

    expect(window.location.pathname).toBe(`/users/${userData.id}`);
  });
});
