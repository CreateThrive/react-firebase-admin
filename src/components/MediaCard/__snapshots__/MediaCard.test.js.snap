// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaCard /> rendering should render without crashing 1`] = `
<DocumentFragment>
  <div
    class="media"
    data-testid="media"
    style="margin-top: auto; margin-bottom: auto; padding-bottom: 1rem;"
  >
    <figure
      class="media-left"
    >
      <p
        class="image is-64x64"
      >
        <img
          alt="Profile logo preview"
          class="is-rounded"
          src="some logoUrl"
        />
      </p>
    </figure>
    <div
      class="media-content"
    >
      <div
        class="content"
      >
        <p
          class="media-meta"
        >
          <strong>
            Mateo
          </strong>
           
          <small
            class="has-text-grey"
          >
            Thu, Nov 12, 2020
          </small>
        </p>
        <p>
          mkrukuy@gmail.com
        </p>
      </div>
    </div>
    <div
      class="media-right"
    >
      <div
        class="field is-grouped is-grouped-multiline"
      >
        <div
          class="control"
        >
          <a
            href="/users/testId"
          >
            <span
              class="icon"
            >
              <i
                class="mdi mdi-account-edit"
              />
            </span>
          </a>
        </div>
        <div
          class="control"
        >
          <span
            class="icon has-text-danger"
            data-testid="delete"
            style="cursor: pointer;"
          >
            <i
              class="mdi mdi-cancel"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
